Bluetooth: controller: nRF53x: Fix missing sw_switch clear DPPI config

cvinayak opened the pull request:
Fix missing sw_switch timer clear DPPI config when
re-enabling Tx or Rx after radio_disable() or
radio_switch_complete_and_disable() call in LLL state/role
contexts.

Fixes #28471.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak commented:
@ioannisg do you know why resetting the sw switch subscription before asserting radio disable task cause instability?

ioannisg commented:
> @ioannisg do you know why resetting the sw switch subscription before asserting radio disable task cause instability?

Need to check this @cvinayak , will come back to this.

cvinayak commented:
@ioannisg Could you approve this?

cvinayak commented:
@ioannisg poke

ioannisg commented:
> @ioannisg poke

Taking a look now, sorry for the huge delay

ioannisg commented:
@cvinayak I remember I had a comment in the #26019 about changing the behavior of the _cleanup function.

ioannisg commented:
> @ioannisg do you know why resetting the sw switch subscription before asserting radio disable task cause instability?

@cvinayak I did some preliminary investigation on this, and I admit the issue seems complex.
The subscribe tasks related to SW Switch are
- TIMER CLEAR
- Group Task Enable

the first task is configured by radio_tmr_start(), so I guess, this occurs once in the event,
the second task is configured at sw_switch() and re-configured each time there's a switch in radio events

the tasks are cleared in radio_switch_complete_and_disable()

#26019 introduced a clearing of these tasks also in radio_disable() and this seems to be causing the issue reported in the bug ticket.
(I remember I 've been asking about the above clearing in that PR Review).

I think that unless this is a weird race condition, de-registering the tasks at radio_disable() should not harm, since the PPIs are also disabled, so any subscription would not matter. That should be the case, as long as nothing re-enables the radio after radio_disable() is called in an event..

ioannisg commented:
If we agree that this is the fix we need, I think it is cleaner that we just revert the offending commit rather than merging this PR. Cleanup does not give any value if we cannot have the same code in both 51/52 and 53 series, and the original code was simpler.

cvinayak commented:
> the tasks are cleared in radio_switch_complete_and_disable()
> 
> #26019 introduced a clearing of these tasks also in radio_disable() and this seems to be causing the issue reported in the bug ticket.
(I remember I 've been asking about the above clearing in that PR Review).

radio_switch_complete_and_disable() will not be called when BLE radio event is pre-empted by call to `radio_disable`, in this case the tasks will not be cleared. This works fine when Zephyr controller is only the user of Radio between events, as these task assignments are probably retained due to Errata 16 (refer to https://github.com/zephyrproject-rtos/zephyr/pull/26018).

There seem to be something wrong in the original implementation which disallows the clearing of tasks by the offending call introduced in #26019. As long as Errata 16 exists in nRF5340, the reverting of the single line in the current PR puts the implementation back to original behavior.

> If we agree that this is the fix we need, I think it is cleaner that we just revert the offending commit rather than merging this PR. Cleanup does not give any value if we cannot have the same code in both 51/52 and 53 series, and the original code was simpler.

The #26019 addresses the use of `hal_radio_nrf_ppi_*` interface which refers to PPI, but then calls `nrf_dppi_*` functions. Original code is very confusing, and difficult to maintain with the level of macros used to hide the executed code. 

If DPPIs are going to be used in more variant, it is better to refactor out the used conditional compilation and use plain C functions instead of macros with code in them.

This refactoring can be addressed in #26118.

ioannisg commented:
> There seem to be something wrong in the original implementation which disallows the clearing of tasks by the offending call introduced in #26019. As long as DPPI is not used by other users between Radio event, the reverting of the single line in the current PR puts the implementation back to original behavior.

If you suspect that something is wrong with the original implementation, I do not understand why we should be reverting to it.

ioannisg commented:
> radio_switch_complete_and_disable() will not be called when BLE radio event is pre-empted by call to radio_disable, in this case the tasks will not be cleared. This works fine when Zephyr controller is only the user of Radio between events, as these task assignments are probably retained due to Errata 16 (refer to #26018), DDPI peripheral retaining the assigned values.

I don't understand this either. radio_disable() will disable the DPPI channel, anyways: there is no issues with any other Radio driver that will preempt the LE Controller; _that_ driver is responsible for re-configuring DPPI and not rely on LE COntroller.

LE Controller also, should not rely on any retained values of DDPI register tasks, when it starts a new radio event.

cvinayak commented:
> If you suspect that something is wrong with the original implementation, I do not understand why we should be reverting to it.

The original implementation does not clear the `HAL_SW_SWITCH_TIMER_CLEAR_PPI_REGISTER_TASK` when `radio_disable` is used; any other Radio user between BLE events will lead to the SWITCH_TIMER being cleared if the DPPI is enabled by the other radio user. This is my concern.

And clearly clearing `HAL_SW_SWITCH_TIMER_CLEAR_PPI_REGISTER_TASK` in `radio_disable` makes the controller unstable when establishing connection as a central.

I think, I may have found the problem in `radio_tmr_start_tick` and `radio_tmr_start_now` interfaces used by scanner... I will update this PR.

cvinayak committed:
Bluetooth: controller: nRF53x: Fix missing sw_switch clear DPPI config

Fix missing sw_switch timer clear DPPI config when
re-enabling Tx or Rx after radio_disable() or
radio_switch_complete_and_disable() call in LLL state/role
contexts.

Fixes #28471.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak commented:
@rick1082 Could you please re-check if this updated PR fixes the central connection issue for you?

rick1082 commented:
@cvinayak I did the following test:

- Program central_hr on nRF5340-DK and peripheral_hr on nRF52832-DK, and **it works properly without issue**.
- Program central_hr on nRF52832-DK and peripheral_hr on nRF5340-DK, and there's a chance that nRF52832-DK cannot connect to peripheral immediately, nRF52832-DK disconnected immediately with reason 0x3e. Not sure if it is related with this ticket.

After multiple trials, I found that:
1. For nRF52832-DK as central_hr, and nRF52840-DK as peripheral_hr, it is easy to establish connection.
2. For nRF52832-DK/nRF52840-DK as central_hr, and nRF5340-DK as peripheral_hr, the central needs to try to connect multiple times (> 5 secs)for establishing the connection. It is easy to replicate this by pushing reset button on nRF5340-DK and let central scanning and try to connect again, since I found central is easier to connect with peripheral at first time, but hard to connect again.

cvinayak commented:
@rick1082 

> Program central_hr on nRF52832-DK and peripheral_hr on nRF5340-DK, and there's a chance that nRF52832-DK cannot connect to peripheral immediately, nRF52832-DK disconnected immediately with reason 0x3e. Not sure if it is related with this ticket.

Do you know the disconnect reason in peripheral_hr on nRF5340-DK? Based on the reason, may be its a different issue, which I can further analyze offline with you.

> I found central is easier to connect with peripheral at first time, but hard to connect again.

This could be different issue, again we can analyze offline.

cvinayak commented:
> I think, I may have found the problem in radio_tmr_start_tick and radio_tmr_start_now interfaces used by scanner... I will update this PR.

I prepared a revert branch for #26019, but then I found some thing else, that `radio_switch_complete_and_disable` unregistered the task after responding with SCAN_RSP on channel 37; which means, channel 38 and channel 39 would not do the sw switch without the `hal_sw_switch_timer_clear_ppi_config` call before `radio_tx_enable`.

Hence, the fix here cf862cb49f763c6cfc11b113c3bbee315f94edaf addresses the regression caused by cleanup function and also addresses already existing unregister happening due to calls to `radio_switch_complete_and_disable`.

ioannisg commented on subsys/bluetooth/controller/ll_sw/nordic/hal/nrf5/radio/radio.c:
```diff
@@ -739,6 +751,9 @@ uint32_t radio_tmr_start_tick(uint8_t trx, uint32_t tick)
 	last_pdu_end_us = 0U;
 #endif /* CONFIG_BT_CTLR_SW_SWITCH_SINGLE_TIMER */
+#if defined(CONFIG_SOC_SERIES_NRF53X)
+	hal_sw_switch_timer_clear_ppi_config();
```
Could you please add an inline comment explaining why this is not needed for nRF51-52?

ioannisg commented:
@kruithofa  @cvinayak now I think I understand the problem, based on your input:

For nRF5340:
- radio_switch_complete_and_disable _originally_ was only clearing the DPPI event/task publish/subscribe, leaving the channel ON (for other events that were subscribed to RADIO END, which is actually needed.)
- radio_disable() _originally_ was disabling the DPPI channel , leaving the publish/subscribe task "dirty".

The refactor commit added the event/task register clearing _also_ in radio_disable().

This seemed to be OK to me, as I had the impression that after radio_disable(), nothing can occur in the current radio event, so it is OK to clear the registers for event publish and task subscribe.

But your input suggests that this is not true; after radio_disable() the radio might need to turn on via a code path that does _not_ re-set the event/task registers for sw-switch.

Your last commit fixes this by adding this "re-set" in the corresponding functions.

Now, I am thinking: could we then just keep fd51199b43b2b93073c8dc116feae09d7f745237 and the removal of hal_radio_sw_switch_disable from nrf53-version of hal_radio_sw_switch_cleanup()? 
AND properly document the reason?

If the result is the same, I would prefer it, since it removes rather than add code.

cvinayak commented:
> Now, I am thinking: could we then just keep [fd51199](https://github.com/zephyrproject-rtos/zephyr/commit/fd51199b43b2b93073c8dc116feae09d7f745237) and the removal of hal_radio_sw_switch_disable from nrf53-version of hal_radio_sw_switch_cleanup()?
> AND properly document the reason?
> 
> If the result is the same, I would prefer it, since it removes rather than add code.

I have added in-line comments in the changes in this PR, which should clarify that fixes here are needed due to the fact that existing use of radio_switch_complete_and_disable() caused timer clear task being unsubscribed. The fix addresses both radio_disable() and radio_switch_complete_and_disable()  calls inside a radio event.

kruithofa reviewed (approved):
LGTM
